### PR TITLE
Enable cluster table functions for DataLake Storages

### LINKED ISSUE #70174: Cluster versions for data lake engines
Every data lake engine, such as `Iceberg`, `DeltaLake`, and optionally, `Hudi`, and the corresponding table functions, `iceberg`, `deltaLake`, and optionally, `hudi`, should have a corresponding `Cluster` versions, such as `IcebergCluster`, etc., similarly to `s3Cluster`, `azureBlobStorageCluster`.

Data lake engines use their metadata and the query to select a bunch of files on object storage and read from them. The parallelization across the cluster will be performed over this bunch of files, similarly how it works for `s3Cluster`.

**Additional Context**

1. We also want to make the use of cluster implicit when parallel replicas are enabled, see another task, https://github.com/ClickHouse/ClickHouse/issues/65024
2. Distribution over the cluster is currently suboptimal, when the number of files is comparable to the number of servers - in this case one server could acquire disproportionally many files for processing. This has to be improved.
3. In the case of a small number of files, we should be able to use a range in a file as a parallelization unit instead of a whole file. This has to be improved.

### PR DESCRIPTION
**Pull request by Michicosun**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add `iceberg[S3;HDFS;Azure]Cluster`, `deltaLakeCluster`, `hudiCluster` table functions. 

Tests added only for iceberg.

Closes: https://github.com/ClickHouse/ClickHouse/issues/70174

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit a29ded494186165469ab207c5fdbc9edb31d9c2a with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/72045/a29ded494186165469ab207c5fdbc9edb31d9c2a/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/72045/a29ded494186165469ab207c5fdbc9edb31d9c2a/style_check.html">❌ failure</a></td></tr>
<tbody>
</table>

File: src/TableFunctions/TableFunctionObjectStorage.cpp [hunk 1]
```diff
@@ -226,6 +226,27 @@ template class TableFunctionObjectStorage<HDFSClusterDefinition, StorageHDFSConf
...
+#if USE_AVRO && USE_AWS_S3
+template class TableFunctionObjectStorage<IcebergClusterDefinition, StorageS3IcebergConfiguration>;
+template class TableFunctionObjectStorage<IcebergS3ClusterDefinition, StorageS3IcebergConfiguration>;
+#endif
+
+#if USE_AVRO && USE_AZURE_BLOB_STORAGE
+template class TableFunctionObjectStorage<IcebergAzureClusterDefinition, StorageAzureIcebergConfiguration>;
+#endif
+
+#if USE_AVRO &&  USE_HDFS
```
Comment (SmitaRKulkarni) [inline src/TableFunctions/TableFunctionObjectStorage.cpp:238]:
```suggestion
#if USE_AVRO && USE_HDFS
```

File: src/TableFunctions/TableFunctionObjectStorageCluster.cpp [hunk 2]
```diff
@@ -105,15 +105,79 @@ void registerTableFunctionObjectStorageCluster(TableFunctionFactory & factory)
...
+         .allow_readonly = false});
 #endif
 
-#if USE_AZURE_BLOB_STORAGE
-template class TableFunctionObjectStorageCluster<AzureClusterDefinition, StorageAzureConfiguration>;
+#if USE_AVRO && USE_AZURE_BLOB_STORAGE
+    factory.registerFunction<TableFunctionIcebergAzureCluster>(
+        {.documentation
+         = {.description = R"(The table function can be used to read the Iceberg table stored on Azure object store in parallel for many nodes in a specified cluster.)",
+            .examples{{"icebergAzureCluster", "SELECT * FROM icebergAzureCluster(url, access_key_id, secret_access_key)", ""}},
```
Comment (SmitaRKulkarni) [inline src/TableFunctions/TableFunctionObjectStorageCluster.cpp:133]:
```suggestion
            .examples{{"icebergAzureCluster", "SELECT * FROM icebergAzureCluster(storage_account_url, blob_path, format)", ""}},
```
Or other azure function signatures

File: src/TableFunctions/TableFunctionObjectStorageCluster.h [hunk 3]
```diff
@@ -33,6 +33,42 @@ struct HDFSClusterDefinition
     static constexpr auto storage_type_name = "HDFSCluster";
 };
 
+struct IcebergClusterDefinition
+{
+    static constexpr auto name = "icebergCluster";
+    static constexpr auto storage_type_name = "IcebergS3Cluster";
```
Comment (SmitaRKulkarni) [inline src/TableFunctions/TableFunctionObjectStorageCluster.h:39]:
```suggestion
    static constexpr auto storage_type_name = "IcebergCluster";
```

File: src/TableFunctions/TableFunctionObjectStorageCluster.cpp [hunk 4]
```diff
@@ -105,15 +105,79 @@ void registerTableFunctionObjectStorageCluster(TableFunctionFactory & factory)
...
+    factory.registerFunction<TableFunctionHudiCluster>(
+        {.documentation
+         = {.description = R"(The table function can be used to read the Hudi table stored on object store in parallel for many nodes in a specified cluster.)",
+            .examples{{"hudiCluster", "SELECT * FROM hudiCluster(url, access_key_id, secret_access_key)", ""}},
+            .categories{"DataLake"}},
+         .allow_readonly = false});
+#endif
+}
+
+void registerDataLakeClusterTableFunctions(TableFunctionFactory & factory)
```
Comment (SmitaRKulkarni) [inline src/TableFunctions/TableFunctionObjectStorageCluster.cpp:176]: Just a suggestion, maybe we can move the checks here and get rid of `maybe_unused` & `UNUSED` in functions, like we do for `registerDataLakeTableFunction` (https://github.com/ClickHouse/ClickHouse/blob/master/src/TableFunctions/TableFunctionObjectStorage.cpp#L298)

**Review by SmitaRKulkarni — COMMENTED** [review]
Rest all LGTM

File: src/TableFunctions/TableFunctionObjectStorageCluster.h (hunk 3 above)
Comment (Michicosun) [reply src/TableFunctions/TableFunctionObjectStorageCluster.h:39]: Did the same as here https://github.com/ClickHouse/ClickHouse/blob/master/src/TableFunctions/TableFunctionObjectStorage.h#L68, do you think it's worth changing?

File: src/TableFunctions/TableFunctionObjectStorage.cpp (hunk 1 above)
Comment (Michicosun) [reply src/TableFunctions/TableFunctionObjectStorage.cpp]: fixed

File: src/TableFunctions/TableFunctionObjectStorageCluster.cpp (hunk 4 above)
Comment (Michicosun) [reply src/TableFunctions/TableFunctionObjectStorageCluster.cpp:174]: done

File: src/TableFunctions/TableFunctionObjectStorageCluster.h (hunk 3 above)
Comment (SmitaRKulkarni) [reply src/TableFunctions/TableFunctionObjectStorageCluster.h:39]: On second thought, do we need `IcebergCluster` function? 
I believe we have iceberg & icebergS3 as we originally just had iceberg which was for S3 then extended it for other object storages. (iceberg is alias of icebergs3) 
what do you think?

Comment (Michicosun) [reply src/TableFunctions/TableFunctionObjectStorageCluster.h:39]: Seems so, let's just keep IcebergS3Cluster.

**Review by SmitaRKulkarni — APPROVED** [review]

**Pull request merged**